File: agents/src/orchestrator.py
from google.adk import LlmAgent, AgentAsTool
from .researcher import get_researcher_agent
from .writer import get_writer_agent
from .publisher import get_publisher_agent

def get_orchestrator():
    researcher = get_researcher_agent()
    writer = get_writer_agent()
    publisher = get_publisher_agent()
    
    return LlmAgent(
        model="gemini-1.5-pro",
        tools=[
            AgentAsTool(researcher, name="researcher"),
            AgentAsTool(writer, name="writer"),
            AgentAsTool(publisher, name="publisher")
        ],
        instructions="""
        Coordinate the LinkedIn article generation and publishing process.
        1. Call 'researcher' with the user's niche and topics.
        2. Pass the brief to 'writer'.
        3. Pass the final article, access_token, and user_urn to 'publisher'.
        
        Ensure you maintain the user's preferences throughout the chain.
        """
    )

File: apps/web/src/agents/orchestrator.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
import { LlmAgent, AgentTool } from '@google/adk';
import { getResearcherAgent } from './researcher';
import { getImageGenAgent } from './imageGen';
import { getWriterAgent } from './writer';
import { getPublisherAgent } from './publisher';

export function getOrchestrator() {
  const researcher = getResearcherAgent();
  const imageGen = getImageGenAgent();
  const writer = getWriterAgent();
  const publisher = getPublisherAgent();
  
  return new LlmAgent({
    name: 'orchestrator',
    model: process.env.MODEL_ID || 'gemini-2.5-flash',
    tools: [
      new AgentTool({ agent: researcher }),
      new AgentTool({ agent: imageGen }),
      new AgentTool({ agent: writer }),
      new AgentTool({ agent: publisher }),
    ],
    instruction: `You are the lead orchestrator for a LinkedIn content generation pipeline. Your mission is to research, design, write, and publish an expert-level article for a user.

Follow this exact sequence:
1. Call the 'researcher' tool with the User ID from the initial message. The researcher will fetch target_niche and target_topics from Firestore users table.
2. Call the 'image_gen' tool with the same niche and topics to gather visual inspiration and generate an image URL. Capture image_gen.image_url.
3. Call the 'writer' tool, providing BOTH the Research Brief and the Image Brief, to draft a high-impact, scannable LinkedIn post.
4. Finally, call the 'publisher' tool. Use the 'access_token' and 'User URN' provided in the initial user message, and pass image_gen.image_url as image_url.

CRITICAL: The 'access_token' is provided in the input data. You MUST pass it exactly as is to the publisher tool.

CRITICAL: If image_gen returns image_url, you MUST pass that exact URL to publisher.image_url.

CRITICAL: Verify the response from 'publisher'. If it fails, report the error. Do not claim success unless the publisher tool confirms the post is live.

FINAL RESPONSE REQUIREMENT:
- You MUST return a final JSON object.
- On success, return: {"status":"published","postId":"...","imageUrl":"...","message":"..."}
- On failure, return: {"status":"failed","error":"..."}
- Never return an empty response.`,
  });
}
